netaddr.go

```diff
@@ -643,22 +651,108 @@ func (ip IP) As4() [4]byte {
 // Note that unlike the Go standard library's IP.String method,
 // IP4-mapped IPv6 addresses do not format as dotted decimals.
 func (ip IP) String() string {
-	if ip.z == z0 {
+	switch ip.z {
+	case z0:
 		return "invalid IP"
+	case z4:
+		return ip.string4()
+	default:
+		return ip.string6()
 	}
-	if ip.Is4() {
-		return fmt.Sprintf("%d.%d.%d.%d", ip.v4(0), ip.v4(1), ip.v4(2), ip.v4(3))
+}
+
+// digits is a string of the hex digits from 0 to f. It's used in
+// appendDecimal and appendHex to format IP addresses.
+const digits = "0123456789abcdef"
+
+// appendDecimal appends the decimal string representation of x to b.
+func appendDecimal(b []byte, x uint8) []byte {
+	// Using this function rather than strconv.AppendUint makes IPv4
+	// string building 2x faster.
+
+	if x >= 100 {
+		b = append(b, digits[x/100])
 	}
-	if ip.Is4in6() {
-		zone := ip.Zone()
-		var percent string
-		if zone != "" {
-			percent = "%"
+	if x >= 10 {
+		b = append(b, digits[x/10%10])
+	}
+	return append(b, digits[x%10])
+}
+
+// appendHex appends the hex string representation of x to b.
+func appendHex(b []byte, x uint16) []byte {
+	// Using this function rather than strconv.AppendUint makes IPv6
+	// string building 2x faster.
+
+	if x >= 0x1000 {
+		b = append(b, digits[x>>12])
+	}
+	if x >= 0x100 {
+		b = append(b, digits[x>>8&0xf])
+	}
+	if x >= 0x10 {
+		b = append(b, digits[x>>4&0xf])
+	}
+	return append(b, digits[x&0xf])
+}
+
+func (ip IP) string4() string {
+	const max = 4*3 + 3 // 4 fields, each 3 decimal digits wide, plus 3 dot separators.
+	ret := make([]byte, 0, max)
+
+	ret = appendDecimal(ret, ip.v4(0))
+	ret = append(ret, '.')
+	ret = appendDecimal(ret, ip.v4(1))
+	ret = append(ret, '.')
+	ret = appendDecimal(ret, ip.v4(2))
+	ret = append(ret, '.')
+	ret = appendDecimal(ret, ip.v4(3))
+
+	return string(ret)
+}
+
+// string6 formats ip in IPv6 textual representation. It follows the
+// guidelines in section 4 of RFC 5952
+// (https://tools.ietf.org/html/rfc5952#section-4): no unnecessary
+// zeros, use :: to elide the longest run of zeros, and don't use ::
+// to compact a single zero field.
+func (ip IP) string6() string {
+
+	zeroStart, zeroEnd := uint8(255), uint8(255)
+	for i := uint8(0); i < 8; i++ {
+		j := i
+		for j < 8 && ip.v6_16(j) == 0 {
+			j++
 		}
-		return fmt.Sprintf("::ffff:%x:%x%s%s", uint16(ip.v4(0))<<8|uint16(ip.v4(1)), uint16(ip.v4(2))<<8|uint16(ip.v4(3)), percent, zone)
```

Comment:
is this a behavior change?

Reply:
No behavior change, the string code no longer treats mapped IPs as special, so they get printed through the generic IPv6 printing code - which formats the same as this line. We had this special case previously to short-circuit the stdlib code, which formats these IPs as `::ffff:1.2.3.4`, but our code never does.

Reply:
ah, I read that fmt.Sprintf too quickly... I was assuming _our_ special case previously was making the `::ffff:1.2.3.4` form. It was the opposite... we were avoiding it. Gotcha.